rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,655 @@
+
+To rectify this, a unique identifier should be established for an instance of
+a Bot.
+
+#### Bot Identity Trust
+
+We should be mindful that adding a new persistent identifier may increase our
+attack surface, particularly if we allow clients manipulate their persistent
+identifier or in any way trust the values communicated to Auth during the join
+process.
+
+For example, trusting this identifier might allow a bot to better masquerade
+itself as a preexisting instance and avoid discovery by an end user that falsely
+assumed no unexpected instances had joined their cluster. Additionally, if we
+implement join limits (i.e. tokens that only allow N bots to join), malicious
+bots could reuse existing identifiers to bypass join limits.
+
+To mitigate this, we should make certain to cryptographically verify identifiers
+during the renewal process. For example, we can embed the identifier as a
+certificate field to ensure it cannot be tampered with once issued by the Auth
+service, or encrypt the renewed certificates using the previous iteration's
+public key to ensure the calling bot owns the private key. Adopting proper mTLS
+during the join process should accomplish both of these goals.
+
+##### Verifying Bot Identities
+
+We currently see two methods for cryptographically verifying bot identities at
+renewal time:
+
+1. We could expose the existing functionality of the HTTPS-only
+   `RegisterUsingToken` over gRPC. The existing gRPC `JoinService` can be
+   accessed with and without authentication, so we could inspect the client
+   connection to find the existing bot identity, if any, and it would be
+   implicitly verified.
+2. We could adapt the existing HTTPS implementation of `RegisterUsingToken` to
+   additionally accept an encoded existing certificate, and return certificates
+   encrypted with the certificate's public key. We can verify the certificate
+   was originally signed with our CA, and the client will only be able to
+   decrypt the returned identity if they actually have the private key for the
+   previous identity.
+
+Our preference is option (1): bots always join over gRPC, and provide a client
+certificate when re-joining. Bots without a client cert to present are
+registered as new instances, while bots with a valid client cert preserve their
+identity.
+
+Note that bots joined with the `token` method are not affected by this, as
+renewals take place over a fully mTLS-authenticated tunnel to the auth service.
+
+Additionally, a downside to certificate verification is that the certificate
+validity period becomes a factor. If bots are only run intermittently, like
+from a CI workflow, their certificates could expire and prevent them from
+being identified as the same instance. This is likely to only impact a small
+number of cases, however, as most CI provider joins are stateless and have no
+certificates to present anyway. Bots that present expired certificates will
```

Comment:
I imagine the TLS handshake will reject them for us - we can just make sure the client is capable of handling this and will try without a certificate.